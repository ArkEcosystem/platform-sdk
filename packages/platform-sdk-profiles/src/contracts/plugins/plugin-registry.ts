export interface IPluginRegistry {
	all(): Promise<IRegistryPlugin[]>;
	size(pkg: IRegistryPlugin): Promise<number>;
	downloads(pkg: IRegistryPlugin): Promise<number>;
}

/* istanbul ignore file */

export type IRegistryPluginAuthor = { [key: string]: any } | string;

export interface IRegistryPluginManifest {
	logo?: string;
	title?: string;
	images?: string[];
	categories?: string[];
	permissions?: string[];
	urls?: any[];
	minVersion?: string;
}

export interface IRegistryPluginSourceControl {
	type: string;
	value: string;
}

export interface IRegistryPluginSocialMedia {
	type: string;
	value: string;
}

export interface IRegistryPluginImage {
	type: string;
	value: string;
}

export interface IRegistryPluginMeta {
	displayName?: string;
	description?: string;
}

export interface RegistryPluginAip36 {
	sourceControl?: IRegistryPluginSourceControl[];
	socialMedia?: IRegistryPluginSocialMedia[];
	images?: IRegistryPluginImage[];
	meta?: IRegistryPluginMeta;
}

export interface IRegistryPluginVersion {
	id: number;
	name: string;
	released_at: Date;
	is_verified: boolean;
}

export interface IRegistryPluginProperties {
	name: string;
	version: string;
	description?: string;
	keywords?: string[];
	homepage?: { [key: string]: any } | string;
	bugs?: { [key: string]: any };
	license?: string;
	author?: IRegistryPluginAuthor;
	contributors?: ({ [key: string]: any } | string)[];
	maintainers?: ({ [key: string]: any } | string)[];
	repository?: { [key: string]: any } | string;
	dist?: {
		shasum?: string;
		tarball?: string;
		unpackedSize?: number;
		[key: string]: any;
	};
	"desktop-wallet"?: IRegistryPluginManifest;
	[key: string]: any;
}

export interface IRegistryPlugin {
	id(): string;
	name(): string;
	alias(): string;
	date(): string;
	version(): string;
	description(): string;
	author(): IRegistryPluginAuthor;
	sourceProvider(): any;
	logo(): string;
	images(): string[];
	categories(): string[];
	permissions(): string[];
	urls(): string[];
	minimumVersion(): string;
	toObject(): {
		id: string;
		name: string;
		alias: string;
		date: string;
		version: string;
		description: string;
		author: IRegistryPluginAuthor;
		sourceProvider: any;
		logo: string;
		images: string[];
		categories: string[];
		permissions: string[];
		urls: string[];
		minimumVersion: string;
	};
}
