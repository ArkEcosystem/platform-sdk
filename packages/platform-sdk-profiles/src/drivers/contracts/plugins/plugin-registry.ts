export interface IPluginRegistry {
    all(): Promise<IRegistryPlugin[]>;
    size(pkg: IRegistryPlugin): Promise<number>;
    downloads(pkg: IRegistryPlugin): Promise<number>;
}

/* istanbul ignore file */

type RegistryPluginAuthor = { [key: string]: any } | string;

export interface RegistryPluginManifest {
	logo?: string;
	title?: string;
	images?: string[];
	categories?: string[];
	permissions?: string[];
	urls?: any[];
	minVersion?: string;
}

export interface RegistryPluginSourceControl {
	type: string;
	value: string;
}

export interface RegistryPluginSocialMedia {
	type: string;
	value: string;
}

export interface RegistryPluginImage {
	type: string;
	value: string;
}

export interface RegistryPluginMeta {
	displayName?: string;
	description?: string;
}

export interface RegistryPluginAip36 {
	sourceControl?: RegistryPluginSourceControl[];
	socialMedia?: RegistryPluginSocialMedia[];
	images?: RegistryPluginImage[];
	meta?: RegistryPluginMeta;
}

export interface RegistryPluginVersion {
	id: number;
	name: string;
	released_at: Date;
	is_verified: boolean;
}

export interface RegistryPluginProperties {
	name: string;
	version: string;
	description?: string;
	keywords?: string[];
	homepage?: { [key: string]: any } | string;
	bugs?: { [key: string]: any };
	license?: string;
	author?: RegistryPluginAuthor;
	contributors?: ({ [key: string]: any } | string)[];
	maintainers?: ({ [key: string]: any } | string)[];
	repository?: { [key: string]: any } | string;
	dist?: {
		shasum?: string;
		tarball?: string;
		unpackedSize?: number;
		[key: string]: any;
	};
	"desktop-wallet"?: RegistryPluginManifest;
	[key: string]: any;
}

export interface IRegistryPlugin {
    id(): string;
    name(): string;
    alias(): string;
    date(): string;
    version(): string;
    description(): string;
    author(): RegistryPluginAuthor;
    sourceProvider(): any;
    logo(): string;
    images(): string[];
    categories(): string[];
    permissions(): string[];
    urls(): string[];
    minimumVersion(): string;
    toObject(): {
		id: string;
		name: string;
		alias: string;
		date: string;
		version: string;
		description: string;
		author: RegistryPluginAuthor;
		sourceProvider: any;
		logo: string;
		images: string[];
		categories: string[];
		permissions: string[];
		urls: string[];
		minimumVersion: string;
	};
}
