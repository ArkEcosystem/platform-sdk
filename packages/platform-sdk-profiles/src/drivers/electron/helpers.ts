import { container } from "../../environment/container";
import { Identifiers } from "../../environment/container.models";
import { ProcessRender } from "./contracts";

export const communicate = async <T = any>(event: string, data: any): Promise<T> => container.get<ProcessRender>(Identifiers.ProcessRenderer).invoke(event, data);
