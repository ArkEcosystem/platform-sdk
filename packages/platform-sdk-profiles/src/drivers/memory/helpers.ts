import EventEmitter from "eventemitter3";

import { container } from "../../environment/container";
import { Events, Identifiers } from "../../environment/container.models";
import { State } from "../../environment/state";

export const emitter = (): EventEmitter => container.get(Identifiers.EventEmitter);

export const emitProfileChanged = (): boolean => emitter().emit(Events.ProfileChanged, { id: State.profile().id() })
