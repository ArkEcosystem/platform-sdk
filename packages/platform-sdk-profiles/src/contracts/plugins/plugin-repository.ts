import { Except } from "type-fest";
import { IPluginRegistry } from "./plugin-registry";

export interface IPlugin {
	id: string;
	name: string;
	version: string;
	isEnabled: boolean;
	permissions: string[];
	urls: string[];
}

export interface IPluginRepository {
    all(): Record<string, IPlugin>;
    first(): IPlugin;
    last(): IPlugin;
    keys(): string[];
    values(): IPlugin[];
    push(plugin: Except<IPlugin, "id">): IPlugin;
    fill(data: object): void;
    findById(id: string): IPlugin;
    forget(id: string): void;
    flush(): void;
    count(): number;
    registry(): IPluginRegistry;
}
