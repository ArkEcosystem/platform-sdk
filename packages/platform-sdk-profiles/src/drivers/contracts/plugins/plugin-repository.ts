import { Except } from "type-fest";
import { IPluginRegistry } from "./plugin-registry";

interface Plugin {
	id: string;
	name: string;
	version: string;
	isEnabled: boolean;
	permissions: string[];
	urls: string[];
}

export interface IPluginRepository {
    all(): Record<string, Plugin>;
    first(): Plugin;
    last(): Plugin;
    keys(): string[];
    values(): Plugin[];
    push(plugin: Except<Plugin, "id">): Plugin;
    fill(data: object): void;
    findById(id: string): Plugin;
    forget(id: string): void;
    flush(): void;
    count(): number;
    registry(): IPluginRegistry;
}
