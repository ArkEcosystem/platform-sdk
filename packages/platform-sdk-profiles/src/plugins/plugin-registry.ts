/* istanbul ignore file */

import { Contracts } from "@arkecosystem/platform-sdk";

import { container } from "../environment/container";
import { Identifiers } from "../environment/container.models";
import { RegistryPlugin } from "./plugin-registry.models";

export class PluginRegistry {
	readonly #httpClient: Contracts.HttpClient;

	public constructor() {
		this.#httpClient = container.get<Contracts.HttpClient>(Identifiers.HttpClient);
	}

	public async all(): Promise<RegistryPlugin[]> {
		const results: Promise<RegistryPlugin>[] = [];

		let i = 0;
		// eslint-disable-next-line no-constant-condition
		while (true) {
			const { objects } = (
				await this.#httpClient.get("https://registry.npmjs.com/-/v1/search", {
					text: "keywords:" + ["@arkecosystem", "desktop-wallet", "plugin"].join(" "),
					from: i * 250,
					size: 250,
					t: +new Date(),
				})
			).json();

			if (objects === undefined) {
				break;
			}

			if (objects && objects.length === 0) {
				break;
			}

			for (const item of objects) {
				results.push(this.findById(item.package.name, item.package.date));
			}

			i++;
		}

		return Promise.all(results);
	}

	private async findById(id: string, date: string): Promise<RegistryPlugin> {
		const [details, downloads] = await Promise.all([
			await this.#httpClient.get(`https://registry.npmjs.com/${id}`),
			await this.#httpClient.get(
				`https://api.npmjs.org/downloads/range/2005-01-01:${new Date().getFullYear() + 1}-01-01/${id}`,
			),
		]);

		return new RegistryPlugin(details.json(), downloads.json().downloads, date);
	}
}
