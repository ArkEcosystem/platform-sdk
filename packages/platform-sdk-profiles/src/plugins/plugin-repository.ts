import { DataRepository } from "../repositories/data-repository";
import { PluginRegistry } from "./plugin-registry";

interface Plugin {
	id: string;
	name: string;
	version: string;
	isEnabled: boolean;
	permissions: string[];
	urls: string[];
}

export class PluginRepository {
	readonly #data: DataRepository;
	readonly #registry: PluginRegistry;

	public constructor() {
		this.#data = new DataRepository();
		this.#registry = new PluginRegistry();
	}

	public all(): Record<string, Plugin> {
		return this.#data.all() as Record<string, Plugin>;
	}

	public first(): Plugin {
		return this.#data.first();
	}

	public last(): Plugin {
		return this.#data.last();
	}

	public keys(): string[] {
		return this.#data.keys();
	}

	public values(): Plugin[] {
		return this.#data.values();
	}

	public push(plugin: Plugin): void {
		this.#data.set(`${plugin.id}`, plugin);
	}

	public fill(data: object): void {
		this.#data.fill(data);
	}

	public findById(id: string): Plugin {
		const plugin: Plugin | undefined = this.#data.get(id);

		if (!plugin) {
			throw new Error(`Failed to find a plugin for [${id}].`);
		}

		return plugin;
	}

	public forget(id: string): void {
		this.#data.forget(id);
	}

	public flush(): void {
		this.#data.flush();
	}

	public count(): number {
		return this.keys().length;
	}

	public registry(): PluginRegistry {
		return this.#registry;
	}
}
