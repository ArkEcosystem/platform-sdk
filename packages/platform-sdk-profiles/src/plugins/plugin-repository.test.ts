import "jest-extended";

import { bootContainer } from "../../test/helpers";
import { PluginRegistry } from "./plugin-registry";
import { PluginRepository } from "./plugin-repository";

const stubPlugin = {
	id: "@hello/world",
	name: "@hello/world",
	version: "1.0.0",
	isEnabled: true,
	permissions: ["something"],
	urls: ["https://google.com"],
};

let subject: PluginRepository;

beforeAll(() => bootContainer());

beforeEach(() => {
	subject = new PluginRepository();
});

it("should return all data", () => {
	expect(subject.all()).toBeObject();
});

it("should return the first item", () => {
	expect(subject.first()).toMatchInlineSnapshot(`undefined`);
});

it("should return the last item", () => {
	expect(subject.last()).toMatchInlineSnapshot(`undefined`);
});

it("should return all data keys", () => {
	expect(subject.keys()).toBeArray();
});

it("should return all data values", () => {
	expect(subject.values()).toBeArray();
});

it("should find a plugin by its ID", () => {
	const uuid: string = subject.push(stubPlugin);

	expect(subject.findById(uuid)).toEqual(stubPlugin);
});

it("should throw if a plugin cannot be found by its ID", () => {
	expect(() => subject.findById(stubPlugin.id)).toThrow(`Failed to find a plugin for [${stubPlugin.id}].`);
});

it("should restore previously created data", () => {
	subject.fill({ [stubPlugin.id]: stubPlugin });

	expect(subject.findById(stubPlugin.id)).toEqual(stubPlugin);
});

it("should forget specific data", () => {
	const uuid: string = subject.push(stubPlugin);

	expect(subject.count()).toBe(1);

	subject.forget(uuid);

	expect(subject.count()).toBe(0);
});

it("should flush the data", () => {
	subject.push(stubPlugin);

	expect(subject.count()).toBe(1);

	expect(subject.flush()).toBeUndefined();

	expect(subject.count()).toBe(0);
});

it("should count the data", () => {
	expect(subject.count()).toBe(0);
});

it("should access the plugin registry", () => {
	expect(subject.registry()).toBeInstanceOf(PluginRegistry);
});
