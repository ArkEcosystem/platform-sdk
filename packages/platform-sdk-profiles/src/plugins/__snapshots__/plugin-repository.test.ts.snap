// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should restore the blacklist 1`] = `
Set {
  1,
}
`;
